neo/Network/P2P/Connection.cs

```diff
@@ -66,18 +68,38 @@ private void WsReceive()
                 failure: ex => new Tcp.ErrorClosed(ex.Message));
         }
 
-        public void Disconnect(bool abort = false)
+        protected void Disconnect(DisconnectReason reason, byte[] data = null)
         {
             disconnected = true;
+
+            var payload = DisconnectPayload.Create(reason, data);
+            var disconnectMessage = Message.Create(MessageCommand.Disconnect, payload);
```

Comment:
We shouldn't send messages in `Connection`. Application layer protocol should be moved to `RemoteNode`.

Reply:
gotcha, fixed.  
But the `Peer.cs` file is a little hard to refactor.

Reply:
This is not easy, how is it going?

Reply:
done